src/lsm/manifest_level.zig

```diff
@@ -388,6 +419,164 @@ pub fn ManifestLevelType(
+        /// Returns the smallest visible range across level A and B that overlaps key_min/max.
+        ///
+        /// For example, for a table in level 2, count how many tables overlap in level 3, and
+        /// determine the span of their entire key range, which may be broader or narrower.
+        ///
+        /// The range.table_count includes the input table from level A represented by key_min/max.
+        /// Thus range.table_count=1 means that the table may be moved directly between levels.
+        ///
+        /// The range keys are guaranteed to encompass all the relevant level A and level B tables:
+        ///   range.key_min = min(a.key_min, b.key_min)
+        ///   range.key_max = max(a.key_max, b.key_max)
+        ///
+        /// This last invariant is critical to ensuring that tombstones are dropped correctly.
+        pub fn compaction_range(
+            level: *const Self,
+            key_min: Key,
+            key_max: Key,
+        ) CompactionRange {
+            var range = CompactionRange{ .table_count = 1, .key_min = key_min, .key_max = key_max, .tables = .{ .buffer = undefined } };
+            const snapshots = [_]u64{lsm.snapshot_latest};
+            var it = level.iterator(
+                .visible,
+                &snapshots,
+                .ascending,
+                KeyRange{ .key_min = range.key_min, .key_max = range.key_max },
+            );
+
+            while (it.next()) |table| : (range.table_count += 1) {
+                assert(table.visible(lsm.snapshot_latest));
+                assert(compare_keys(table.key_min, table.key_max) != .gt);
+                assert(compare_keys(table.key_max, range.key_min) != .lt);
+                assert(compare_keys(table.key_min, range.key_max) != .gt);
+
+                // The first iterated table.key_min/max may overlap range.key_min/max entirely.
+                if (compare_keys(table.key_min, range.key_min) == .lt) {
+                    range.key_min = table.key_min;
+                }
+
+                // Thereafter, iterated tables may/may not extend the range in ascending order.
+                if (compare_keys(table.key_max, range.key_max) == .gt) {
+                    range.key_max = table.key_max;
+                }
+                // This const cast is safe as we know that the memory pointed to is in fact
+                // mutable. That is, the table is not in the .text or .rodata section.
+                // We set the capacity of range.tables to lsm_growth_factor, since that is the
+                // maximum number of tables that can intersect with a table in Level A. We needn't
+                // add any more table pointers to this array, since this table won't get selected
+                // if it intersects with > lsm_growth_factor number of tables.
+                if (range.tables.len < range.tables.capacity()) {
+                    range.tables.appendAssumeCapacity(@intToPtr(*TableInfo, @ptrToInt(table)));
+                }
```

Comment:
As per my understanding: since [ManifestLevel.compaction_range](https://github.com/tigerbeetledb/tigerbeetle/blob/1b64cac51bc103db6c673ad59dc69d7c12f04e6e/src/lsm/manifest_level.zig#L534) iterates through all tables in Level B that intersect with a given table in Level A. That count could be more that `lsm_growth_factor` (even though we wouldn't end up choosing that as the optimal compaction table), right? I might be missing some context here. 😄 

Also, if my understanding of the above is correct, then I could add a small change that exits this function as soon as we find that a table in Level A intersects with more than `lsm_growth_factor` number of tables.

Reply:
Gotcha. Could we return a different value if we need to append a table but the array is already full? I think its misleading to return what is technically an incomplete range, even if we are guaranteed that the caller won't actually use it.

Reply:
Agreed, returning an incomplete range does seem misleading! 

Shall we add an argument called `max_intersecting_tables`/`intersection_threshold` to ManifestLevel.compaction_range and change its return type to `?CompactionRange`? `max_intersecting_tables` would be set to `lsm_growth_factor`. We could then return `null` if the # of tables that are intersecting cross that threshold. However, we would have to keep `CompactionRange.tables` as an `ArrayList` instead of a `BoundedArray` since the number of elements in the array would be driven by `max_intersecting_tables`/`intersection_threshold`.